artemis-core-client/src/main/java/org/apache/activemq/artemis/core/message/impl/CoreMessage.java

```diff
@@ -52,6 +52,62 @@
  *  consumers */
 public class CoreMessage extends RefCountMessage implements ICoreMessage {
 
+   private static final class CoreTypedProperties extends TypedProperties {
+
+      // We use properties to establish routing context on clustering.
+      // However if the client resends the message after receiving, it needs to be removed
+      private static final Predicate<SimpleString> INTERNAL_PROPERTY_NAMES_CLEANUP_FILTER =
+         name -> (name.startsWith(Message.HDR_ROUTE_TO_IDS) && !name.equals(Message.HDR_ROUTE_TO_IDS)) ||
+            (name.startsWith(Message.HDR_ROUTE_TO_ACK_IDS) && !name.equals(Message.HDR_ROUTE_TO_ACK_IDS));
+      private static final SimpleString AMQ_PROPNAME = new SimpleString("_AMQ_");
+      private boolean internalProperties;
+
+      CoreTypedProperties() {
+         super();
+         internalProperties = false;
+      }
+
+      private CoreTypedProperties(TypedProperties other) {
+         super(other);
+         if (other instanceof CoreTypedProperties) {
+            internalProperties = ((CoreTypedProperties) other).internalProperties;
```

Comment:
CoreTypedProperties is not sync safe due to direct access to  ((CoreTypedProperties) other).internalProperties;

Best solution is to create a private syncronized get method to access internalProperties from the other variable. This then means sync bloc in copyOf method is not needed and this constructor becomes sync safe.